### ISSUE TITLE
tmux pipe-pane from  system()

### ISSUE THREAD

**Issue by iago-lito** [neovim/neovim#9844]
I cannot pipe the content of tmux pane with `pipe-pane` when the tmux session is launched from neovim, whereas I can when it is launched from outside.

- `nvim --version`:
```
NVIM v0.3.4
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -Wconversion -O2 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fdiagnostics-color=auto -Wno-array-bounds -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.3.4/src -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Features: +acl +iconv +jemalloc +tui 
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```
- Operating system/version:
```
$ uname -a
Linux etak 5.0.4-arch1-1-ARCH #1 SMP PREEMPT Sat Mar 23 21:00:33 UTC 2019 x86_64 GNU/Linux
```
- Terminal name/version:
```
$ gnome-terminal --version
# GNOME Terminal 3.32.0 using VTE 0.56.0 +GNUTLS
```
- `$TERM`:
```
$ echo $TERM
xterm-256color
```

### Steps to reproduce using `nvim -u NORC`

```
$ nvim -u NORC # in terminal (0)
:call system('gnome-terminal -- tmux new-session -s TestSession')
# A new tmuxed terminal opens, call it (1).

# Then from another terminal (2), try retrieving the output of (1) with:
$ mkfifo testFifo
$ tmux pipe-pane -s TestSession -o 'cat >> testFifo'
$ cat testFifo # terminal (2) hangs..

# Now type anything into terminal (1)
```

### Actual behaviour

Nothing prints to terminal (2) as input is provided to (1).

### Expected behaviour

(2) should mirror anything happening in (1).
This would be the case if `$ gnome-terminal -- tmux new-session -s TestSession` was directly launched from terminal (0) instead of from within `nvim`.

In a nutshell, the `pipe-pane` works as expected when `tmux new-session` is launched from terminal, but not when launched with vim's `system()`.

What could be causing the problem?  
What difference are there between vim's `system()` execution environment and the plain (0) terminal?

Should I cross-post the issue to [tmux](https://github.com/tmux/tmux)?

**Comment by justinmk** [comment]
Thanks for the clear steps! Just for completeness, instead of `system()` could you  also try:

- `:terminal` in Nvim
- `:!` in Nvim 
- `system()` in gVim (Vim GUI) starting without any configuration (`gvim -N -u NORC`)

> What difference are there between vim's `system()` execution environment and the plain (0) terminal?

At least one difference is that   Nvim always spawns jobs in a new progress group / session: https://github.com/neovim/neovim/blob/53242921987c4641c686f63d7504b7389b4c85d6/src/nvim/event/libuv_process.c#L39-L40

In Vim this is not always the case.

However I wouldn't expect this to affect `tmux new-session -s` because AFAIK tmux looks for sockets per-user, irrespective of job-control things.

More likely this is related to ["pipe" vs "terminal"](https://github.com/neovim/neovim/wiki/FAQ#-and-system-do-weird-things-with-interactive-processes).

**Comment by iago-lito** [comment]
And thank you for reaction :)

I agree that it is weird that tmux seems affected by it being launched one way or the other, but I can confirm the unexpected behaviour even if I launch `new-session` from:
- `:terminal` in neovim
- `:!` in neovim
- `:call system()` in gvim (note that `:echo has('nvim')` yields `0` then)
as you suggested.

Can you reproduce, by the way?

Would reading `nvim -d tp*` help, with the two files being produced by `:call system('set > tp1')` and `$ set > tp2`? There seems to be a little bunch of differences in those two "environment" outputs.

**Comment by iago-lito** [comment]
Wops. This is worse than I first thought. If I launch *another* session from vim with:
`:call system('gnome-terminal -- tmux new-session -s OtherSession')`..

.. and now I try to mirror the naive `$ gnome-terminal -- tmux new-session -s TestSession` directly launched from (0).. then it doesn't work anymore.

The unexpected behaviour therefore seems to propagate accross tmux sessions :\

**Comment by justinmk** [comment]
> Can you reproduce, by the way?

No, because on my machine (tmux 2.8) `-s` option of `tmux pipe-pane -s`  is apparently not supported.


> Would reading `nvim -d tp*` help, with the two files being produced by `:call system('set > tp1')` and `$ set > tp2`? There seems to be a little bunch of differences in those two "environment" outputs.

Depends on how the shell is started. `:terminal` should not have many differences, whereas `system()` starts a non-interactive shell.

**Comment by iago-lito** [comment]
Unfortunately, I'm not sure I understand #1496 :(

It does seem like it's an old, unresolved limitation though.. any workaround for #9844 then?

> In any case, because gVim has the same behavior, this is a duplicate of #1496 .

I'm still not sure I have tried with the right `gvim` application, since `:echo has('nvim')` yielded `0`. Is this what you were expecting?

**Comment by iago-lito** [comment]
Wop: here is the workaround: just prefix your call with `nohup` :)

```
:call system('nohup gnome-terminal -- tmux new-session -s Test')
```

**Comment by justinmk** [comment]
> I'm still not sure I have tried with the right `gvim` application, since `:echo has('nvim')` yielded `0`. Is this what you were expecting?

Yes, the point is to try Vim GUI instead of Vim terminal application.

**Comment by iago-lito** [comment]
Severe update on this: this whole report is irrelevant, for the error simply was that `cat >> testFifo` is a *relative* path specification, and neovim just happened to be launched in another folder. Use `cat >> /absolute/path/to/fifo` instead or watch the local directory you are currently standing in, and everything should be fine. In a nutshell: there is *no* unexpected behaviour.

Sorry for bothering, I wish I could double-close this issue
now :confounded: 

Thank you for support anyway. And thank you for neovim :)

**Comment by justinmk** [comment]
Haha. Thanks for the followup. That makes more sense after all.
